src/frontends/lean/widget.cpp

```diff
@@ -180,6 +188,30 @@ struct stateful_hook : public hook_cell {
     stateful_hook(vm_obj const & init, vm_obj const & pc, vm_obj const & update) : m_init(init), m_props_changed(pc), m_update(update) {}
 };
 
+struct effects_hook : public hook_cell {
+    ts_vm_obj const m_emit;
+    optional<ts_vm_obj> m_props;
+    effects_hook(vm_obj const & emit): m_emit(emit) {}
+    void initialize(vm_obj const & props) override {
+        m_props = optional<ts_vm_obj>(props);
+    }
+    optional<vm_obj> action(vm_obj const & action) override {
+        lean_assert(m_props);
+        vm_obj es = invoke(m_emit.to_vm_obj(), (*m_props).to_vm_obj(), action);
+        get_global_widget_context()->m_effects.push_back(es);
+        return optional<vm_obj>(action);
+    }
+};
+
+LEAN_THREAD_PTR(widget_context, g_context);
+widget_context * get_global_widget_context() {
+    lean_assert(g_context);
+    return g_context;
+}
+void set_global_widget_context(widget_context * wc) {
```

Comment:
Please turn this into a `scope_global_widget` class.

Ideally, I'd like to avoid additional global variables.  Is it instead possible to turn this into an extra argument to `handle_event`?